benchmarking/benchmark.py

```diff
@@ -149,48 +149,77 @@ def upload_data(participant, auth_token, clk_length):
-    mask_a = np.isin(a_ids, b_ids)
-    mask_b = np.isin(b_ids, a_ids)
-    return a_ids, b_ids, mask_a, mask_b
-
-
-def score_mapping(mapping, truth_a, truth_b, mask_a, mask_b):
-    tp = fp = tn = fn = 0
-
-    for idx in range(len(truth_a)):
-        if idx in mapping.keys():
-            if mask_a[idx]:
-                if truth_a[idx] == truth_b[mapping[idx]]:
-                    tp += 1
-                else:
-                    fp += 1
-            else:
-                fp += 1
-        else:
-            if mask_a[idx]:
-                fn += 1
-            else:
-                tn += 1
-    return tp, tn, fp, fn
+        read_values_from_csv(f, 1)
+
+    # Now filter all the entities which are non matched, i.e. they are present only in a single dataset
+    # We filter them because the group results that the entity service is creating does not include them.
+    matched_entities_only = filter(lambda x: len([y for y in groups[x] if y > -1]) > 1, groups)
+    # Finally, map this dictionary into a list of tuples, where the tuples are built from the values in the dictionary
+    # (We drop the key which has no reason anymore).
+    return map(lambda x: tuple(groups[x]), matched_entities_only)
+
+
+def score_accuracy(groups, truth_groups):
+
+    def group_transform(group):
+        """
+        Transform a group provided by the entity service into a tuple of size the number of participants. The ith
+        element of the tuple is equal to the entity's row in the dataset i. The value is -1 if the entity is not in the
+        dataset.
+
+        :param group: List of 2 elements array representing a matched entity where each element (x, y) of the list
+         means that the entity is in the dataset x in the row y.
+        """
+        result = [-1] * 2
+        for link in group:
+            result[link[0]] = link[1]
+        return tuple(result)
+
+    transformed_groups = [group_transform(group) for group in groups]
+
+    # Now everything is in a good form, let's use sets to do an intersection.
+    # Current computed values are:
+    # - positives: all the groups which have been computed by the entity service and which are in the ground truth
+    #       (exact group, not counting parts of groups),
+    # - negatives: number of groups in the ground truth which have not been found by the entity service,
+    # - false_positives: number of groups which have been found by the entity service but shouldn't have been.
+    truth_groups_set = set(truth_groups)
+    positives = len(truth_groups_set.intersection(set(transformed_groups)))
+    negatives = len(truth_groups_set) - positives
+    false_positives = len(transformed_groups) - positives
+    return positives, negatives, false_positives
 
 
 def compose_result(status, tt, experiment, sizes, threshold):
-    tp, tn, fp, fn = tt
+    positives, negatives, false_positives = tt
+    sizes_dic = {'size_{}'.format(chr(97 + i)): sizes[i] for i in range(len(sizes))}
```

Comment:
what's that for? It isn't used anywhere. I guess you eventually will build the 'sizes' section in the `result` with that. Am I right?

Reply:
You're right. I was trying to split the multi-party linkage benchmark update in two PRs, but this is a remaining of a wrong split. It becomes useful in the following PR.